packages/storage-plus/src/snapshot.rs

```diff
@@ -0,0 +1,371 @@
+        let old = self.may_load(store, k.clone())?;
+        self.changelog
+            .save(store, (k, U64Key::from(height)), &ChangeSet { old })
+    }
+
+    pub fn save(&self, store: &mut dyn Storage, k: K, data: &T, height: u64) -> StdResult<()> {
+        if self.should_checkpoint(store, &k)? {
+            self.write_change(store, k.clone(), height)?;
+        }
+        self.primary.save(store, k, data)
+    }
+
+    pub fn remove(&self, store: &mut dyn Storage, k: K, height: u64) -> StdResult<()> {
+        if self.should_checkpoint(store, &k)? {
+            self.write_change(store, k.clone(), height)?;
+        }
+        self.primary.remove(store, k);
+        Ok(())
+    }
+
+    /// load will return an error if no data is set at the given key, or on parse error
+    pub fn load(&self, store: &dyn Storage, k: K) -> StdResult<T> {
+        self.primary.load(store, k)
+    }
+
+    /// may_load will parse the data stored at the key if present, returns Ok(None) if no data there.
+    /// returns an error on issues parsing
+    pub fn may_load(&self, store: &dyn Storage, k: K) -> StdResult<Option<T>> {
+        self.primary.may_load(store, k)
+    }
+
+    // may_load_at_height reads historical data from given checkpoints.
+    // only guaranteed to give correct data if Strategy::EveryBlock or
+    // Strategy::Selected and h is registered as checkpoint
+    pub fn may_load_at_height(
+        &self,
+        store: &dyn Storage,
+        k: K,
+        height: u64,
+    ) -> StdResult<Option<T>> {
+        // this will look for the first snapshot of the given address >= given height
+        // If None, there is no snapshot since that time.
+        let start = Bound::inclusive(U64Key::new(height));
+        let first = self
+            .changelog
+            .prefix(k.clone())
+            .range(store, Some(start), None, Order::Ascending)
+            .next();
+
+        if let Some(r) = first {
+            // if we found a match, return this last one
+            r.map(|(_, v)| v.old)
+        } else {
+            // otherwise, return current value
+            self.may_load(store, k)
+        }
+    }
+
+    /// Loads the data, perform the specified action, and store the result
+    /// in the database. This is shorthand for some common sequences, which may be useful.
+    ///
+    /// If the data exists, `action(Some(value))` is called. Otherwise `action(None)` is called.
+    ///
+    /// This is a bit more customized than needed to only read "old" value 1 time, not 2 per naive approach
+    pub fn update<A, E>(
+        &self,
+        store: &mut dyn Storage,
+        k: K,
+        height: u64,
+        action: A,
+    ) -> Result<T, E>
+    where
+        A: FnOnce(Option<T>) -> Result<T, E>,
+        E: From<StdError>,
+    {
+        let input = self.may_load(store, k.clone())?;
+        let diff = ChangeSet { old: input.clone() };
```

Comment:
This one can be computed inside the `should_checkpoint()` block above.

Reply:
I have to clone the `input` here (as we consume it in `action`). But you are right, `ChangeSet` can be done below.